sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileClient.java

```diff
@@ -1121,12 +1178,63 @@ public ShareFileDownloadResponse downloadWithResponse(OutputStream stream, Share
+                    ResponseBase<FilesDownloadHeaders, InputStream> response = downloadRange(range,
+                        getRangeContentMd5, requestConditions, context);
+                    currentETag = ModelHelper.getETag(response.getHeaders());
+                    if (initialETag != null && !initialETag.equals(currentETag)) {
+                        throw new ConcurrentModificationException("File has been modified concurrently. Expected eTag: "
+                            + initialETag + ", Received eTag: " + currentETag);
+                    }
+                    initialETag = currentETag; // Update eTag for subsequent retries
+
+                    ShareFileDownloadHeaders headers = ModelHelper.transformFileDownloadHeaders(
+                        response.getDeserializedHeaders(), response.getHeaders());
+
+                    try (InputStream responseStream = response.getValue()) {
+                        if (responseStream != null) {
+                            byte[] buffer = new byte[8192];
+                            int bytesRead;
+                            while ((bytesRead = responseStream.read(buffer)) != -1) {
+                                stream.write(buffer, 0, bytesRead);
+                            }
+                        }
+                    }
+                    return new ShareFileDownloadResponse(new ShareFileDownloadAsyncResponse(response.getRequest(),
+                        response.getStatusCode(), response.getHeaders(), null, headers));
+                } catch (IOException e) {
+                    System.out.println("inside the catch clause");
+                    retryCount++;
+                    LOGGER.info("Retrying download due to IOException. Attempt: " + retryCount);
+                } catch (ConcurrentModificationException e) {
+                    throw new ConcurrentModificationException("File has been modified concurrently. Expected eTag: "
+                        + initialETag + ", Received eTag: " + currentETag);
+                }
+            }
+            throw new IllegalStateException("Failed to download file. Max retry attempts reached.");
+        };
 
-        Mono<ShareFileDownloadResponse> download = shareFileAsyncClient.downloadWithResponse(options, context)
-            .flatMap(response -> FluxUtil.writeToOutputStream(response.getValue(), stream)
-                .thenReturn(new ShareFileDownloadResponse(response)));
+        return sendRequest(operation, timeout, ShareStorageException.class);
+    }
 
-        return StorageImplUtils.blockWithOptionalTimeout(download, timeout);
+    private ResponseBase<FilesDownloadHeaders, InputStream> downloadRange(ShareFileRange range,
+        Boolean rangeGetContentMD5, ShareRequestConditions requestConditions, Context context) {
+        String rangeString = range == null ? null : range.toHeaderValue();
+        return azureFileStorageClient.getFiles().downloadWithResponse(shareName, filePath, null,
```

Comment:
There is an option in Autorest which alters `InputStream` responses to `BinaryData`, we may want to look into making that change in the future as `BinaryData` has APIs that write the contents to `OutputStream` and `Channel` which would simplify the code in `downloadToFileWithResponse` and `downloadWithResponse` from stuff like:

```java
try (InputStream responseStream = response.getValue()) {
    if (responseStream != null) {
        byte[] buffer = new byte[8192];
        int bytesRead;
        while ((bytesRead = responseStream.read(buffer)) != -1) {
            stream.write(buffer, 0, bytesRead);
        }
    }
}
```

to

```java
BinaryData streamResponse = response.getValue();
if (streamResponse != null) {
    streamResponse.writeTo(outputStream);
}
```

Reply:
Unfortunately with regenerating the Swagger and using BinaryData, it changes the method signatures for async methods as well. Looks like a bug on the autorest generation.